src/Compilers/CSharp/Portable/SourceGeneration/IncrementalGeneratorInitializationContextExtensions.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Linq;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.SourceGeneration;
+
+internal static partial class IncrementalGeneratorInitializationContextExtensions
+{
+    private static readonly char[] s_nestedTypeNameSeparators = new char[] { '+' };
+
+    /// <summary>
+    /// Returns all syntax nodes of type <typeparamref name="T"/> if that node has an attribute on it that could
+    /// possibly bind to the provided <paramref name="fullyQualifiedMetadataName"/>. <paramref
+    /// name="fullyQualifiedMetadataName"/> should be the full, qualified, metadata name of the attribute, including the
+    /// <c>Attribute</c> suffix.  For example <c>System.CLSCompliantAttribute</c> for <see
+    /// cref="System.CLSCompliantAttribute"/>.
+    /// <para/> This provider understands <see langword="using"/> aliases and will find matches even when the attribute
+    /// references an alias name.  For example, given:
+    /// <code>
+    /// using XAttribute = System.CLSCompliantAttribute;
+    /// [X]
+    /// class C { }
+    /// </code>
+    /// Then
+    /// <c>context.SyntaxProvider.CreateSyntaxProviderForAttribute&lt;ClassDeclarationSyntax&gt;(typeof(CLSCompliantAttribute).FullName)</c>
+    /// will find the <c>C</c> class.
+    /// </summary>
+    public static IncrementalValuesProvider<T> ForAttributeWithMetadataName<T>(this IncrementalGeneratorInitializationContext context, string fullyQualifiedMetadataName)
+        where T : SyntaxNode
+    {
+        var metadataName = fullyQualifiedMetadataName.IndexOf('+') >= 0
+            ? MetadataTypeName.FromFullName(fullyQualifiedMetadataName.Split(s_nestedTypeNameSeparators).Last())
+            : MetadataTypeName.FromFullName(fullyQualifiedMetadataName);
+
+        var simpleTypeName = metadataName.UnmangledTypeName;
+        var nodesWithAttributesMatchingSimpleName = context.SyntaxProvider.CreateSyntaxProviderForAttribute<T>(simpleTypeName);
+
+        var collectedNodes = nodesWithAttributesMatchingSimpleName.Collect().WithTrackingName("collectedNodes_ForAttributeWithMetadataName");
```

Comment:
❗ This is going to cause all nodes to be re-examined even when only one source file changes. It's important to avoid `Collect()` during this entire helper for maximum incrementality.

Reply:
interesting.  any thoughts on this @chsienki .  I technically don't need this, but it does allow me to group all syntax nodes from the same file so taht i can get the semanticmodel for them once.  That will help the admittedly uncommon (but still realistic) case where there is a file with tons of nodes in it taht turn out to be matches.  For example a file containing all the model types for a user with attributes on all of htem.

Reply:
ah ok.  happy to do that.